dagger-core/src/main/java/com/gotocompany/dagger/core/sink/http/HttpSinkWriter.java

```diff
@@ -0,0 +1,268 @@
+    private final Map<String, Function<Row, Object>> customFieldExtractors;
+    private final HttpSinkWriterMetrics metrics;
+    private final HttpSinkWriterState state;
+
+    public HttpSinkWriter(ProtoSerializer protoSerializer, Sink httpSink, int batchSize,
+                          ErrorReporter errorReporter, Set<ErrorType> errorTypesForFailing) {
+        this.protoSerializer = protoSerializer;
+        this.httpSink = httpSink;
+        this.batchSize = batchSize;
+        this.errorReporter = errorReporter;
+        this.errorTypesForFailing = errorTypesForFailing;
+        this.messageQueue = new LinkedBlockingQueue<>(DEFAULT_QUEUE_CAPACITY);
+        this.executorService = Executors.newFixedThreadPool(DEFAULT_THREAD_POOL_SIZE);
+        this.scheduledExecutorService = Executors.newSingleThreadScheduledExecutor();
+        this.currentBatchSize = new AtomicInteger(0);
+        this.customFieldExtractors = initializeCustomFieldExtractors();
+        this.metrics = new HttpSinkWriterMetrics();
+        this.state = new HttpSinkWriterState();
+
+        initializePeriodicFlush();
+    }
+
+    @Override
+    public void write(Row element, Context context) throws IOException, InterruptedException {
+        metrics.incrementTotalRowsReceived();
+        byte[] key = protoSerializer.serializeKey(element);
+        byte[] value = enrichAndSerializeValue(element);
+        Message message = new Message(key, value);
+
+        if (!messageQueue.offer(message, 1, TimeUnit.SECONDS)) {
+            metrics.incrementDroppedMessages();
+            log.warn("Message queue is full. Dropping message: {}", message);
+            return;
+        }
+
+        if (currentBatchSize.incrementAndGet() >= batchSize) {
+            flushQueueAsync();
+        }
+    }
+
+    @Override
+    public List<Void> prepareCommit(boolean flush) throws IOException, InterruptedException {
+        if (flush) {
+            flushQueue();
+        }
+        return Collections.emptyList();
+    }
+
+    @Override
+    public List<Void> snapshotState(long checkpointId) {
+        state.setLastCheckpointId(checkpointId);
+        state.setLastCheckpointTimestamp(System.currentTimeMillis());
+        return Collections.emptyList();
+    }
+
+    @Override
+    public void close() throws Exception {
+        flushQueue();
+        executorService.shutdown();
+        scheduledExecutorService.shutdown();
+        if (!executorService.awaitTermination(30, TimeUnit.SECONDS)) {
+            executorService.shutdownNow();
+        }
+        if (!scheduledExecutorService.awaitTermination(30, TimeUnit.SECONDS)) {
+            scheduledExecutorService.shutdownNow();
+        }
+        httpSink.close();
+    }
+
+    private Map<String, Function<Row, Object>> initializeCustomFieldExtractors() {
+        Map<String, Function<Row, Object>> extractors = new HashMap<>();
+        extractors.put("timestamp", row -> System.currentTimeMillis());
```

Comment:
We are replacing timestamp in row with time of processing. Why we need to do this ?

Reply:
actually later as part of http sink  instrumentation, we want to measure the sink response time , so we need latency between when data enters the system and when it's processed by the sink. this timestamp field is different, the event time will still be retained in event_timestamp field. Also this code will run after the SQL processing is already done, basically after result of SQL query has been obtained